ibtk/src/solvers/impls/BGaussSeidelPreconditioner.cpp

```diff
@@ -182,10 +183,8 @@ BGaussSeidelPreconditioner::solveSystem(SAMRAIVectorReal<NDIM, double>& x, SAMRA
 
     // Clone the right-hand-side vector to avoid modifying it during the
     // preconditioning operation.
-    Pointer<SAMRAIVectorReal<NDIM, double> > f = b.cloneVector(b.getName());
-    f->allocateVectorData();
-    f->copyVector(Pointer<SAMRAIVectorReal<NDIM, double> >(&b, false), false);
-    std::vector<Pointer<SAMRAIVectorReal<NDIM, double> > > f_comps = getComponentVectors(f);
+    ScopedVectorCopy<double> f(b);
+    auto f_comps = f.getComponentVectors();
```

Comment:
@boyceg We can get rid of most of the boilerplate around vector copies with our own vector copy class. If you think this is a good approach I can apply it everywhere.

One possibility would be to use this or a similar class to manage data with RAII - when we destroy an internal class containing these vectors everything will be cleaned up correctly.

Reply:
I like it.

Reply:
Since it's more of a "code cleanup" than bugfix, I'll move that (and I'll add `ScopedVectorDuplicate`) and the patch adding the utility functions to a new PR.

I put all of the 'marker data persisting across a regrid' into a separate internal class of `IBExplicitHierarchyIntegrator`. I think it would be useful for us to do something similar for the classes in this PR so that we can explicitly destroy and recreate objects which need to be reset. One major difficulty in working with these SAMRAI vectors is that we now have three possible states for a vector:
1. has a data index and is allocated
2. has a data index and is not allocated
3. neither has a data index nor is allocated
We can sidestep the 'wrong number of levels' bug and also questions about object state by just recreating them as needed.